#### Open and read the file `Hawking-Questioning-the-Universe.txt`.

In [1]:
file = open("Hawking-Questioning-the-Universe.txt", "r")
f = file.read()
file.close()

#### Print the first 50 characters.

In [3]:
print(f[0:51])

There is nothing bigger or older than the universe.


#### Print from the 52nd to the 163rd character.

In [4]:
print(f[52:164])

The questions I would like to talk about are: one, where did we come from? How did the universe come into being?


---

#### Open the file `Hawking-Questioning-the-Universe.txt` and create a list named `text_as_list` where every word is an element of the list. 

Hint: there are two ways to do it. Either with `.readlines()` or with `.split()`. I encourage you to google how both work, but I suggest to adopt the latter.

Hint: if you want to split the text so every word is an element of a list, what should be the parameter of the `split()` method? Read more [here](https://www.w3schools.com/python/ref_string_split.asp).

In [24]:
with open('Hawking-Questioning-the-Universe.txt') as f:
    text_as_list=[word for line in f for word in line.split()]
print(text_as_list)

['There', 'is', 'nothing', 'bigger', 'or', 'older', 'than', 'the', 'universe.', 'The', 'questions', 'I', 'would', 'like', 'to', 'talk', 'about', 'are:', 'one,', 'where', 'did', 'we', 'come', 'from?', 'How', 'did', 'the', 'universe', 'come', 'into', 'being?', 'Are', 'we', 'alone', 'in', 'the', 'universe?', 'Is', 'there', 'alien', 'life', 'out', 'there?', 'What', 'is', 'the', 'future', 'of', 'the', 'human', 'race?', 'Up', 'until', 'the', '1920s,', 'everyone', 'thought', 'the', 'universe', 'was', 'essentially', 'static', 'and', 'unchanging', 'in', 'time.', 'Then', 'it', 'was', 'discovered', 'that', 'the', 'universe', 'was', 'expanding.', 'Distant', 'galaxies', 'were', 'moving', 'away', 'from', 'us.', 'This', 'meant', 'they', 'must', 'have', 'been', 'closer', 'together', 'in', 'the', 'past.', 'If', 'we', 'extrapolate', 'back,', 'we', 'find', 'we', 'must', 'have', 'all', 'been', 'on', 'top', 'of', 'each', 'other', 'about', '15', 'billion', 'years', 'ago.', 'This', 'was', 'the', 'Big', 'Bang

#### This code will remove the punctuation from your list and applies the lower() method that returns a string where all characters are lower case. Simple run the cell.

In [25]:
import string
text_as_list = [i.translate(str.maketrans('', '', string.punctuation)).lower() for i in text_as_list]
print(text_as_list)

['there', 'is', 'nothing', 'bigger', 'or', 'older', 'than', 'the', 'universe', 'the', 'questions', 'i', 'would', 'like', 'to', 'talk', 'about', 'are', 'one', 'where', 'did', 'we', 'come', 'from', 'how', 'did', 'the', 'universe', 'come', 'into', 'being', 'are', 'we', 'alone', 'in', 'the', 'universe', 'is', 'there', 'alien', 'life', 'out', 'there', 'what', 'is', 'the', 'future', 'of', 'the', 'human', 'race', 'up', 'until', 'the', '1920s', 'everyone', 'thought', 'the', 'universe', 'was', 'essentially', 'static', 'and', 'unchanging', 'in', 'time', 'then', 'it', 'was', 'discovered', 'that', 'the', 'universe', 'was', 'expanding', 'distant', 'galaxies', 'were', 'moving', 'away', 'from', 'us', 'this', 'meant', 'they', 'must', 'have', 'been', 'closer', 'together', 'in', 'the', 'past', 'if', 'we', 'extrapolate', 'back', 'we', 'find', 'we', 'must', 'have', 'all', 'been', 'on', 'top', 'of', 'each', 'other', 'about', '15', 'billion', 'years', 'ago', 'this', 'was', 'the', 'big', 'bang', 'the', 'begi

#### Create a new list `words_begin_with_a` and append all words of the `text_as_list` that begin with a small _a_.

Hint: you have to define first an empty list.

Hint: in order to find if a string starts with one or more letters, you should use the method `.startswith()`. Read more [here](https://www.w3schools.com/python/ref_string_startswith.asp).

In [31]:
words_begin_with_a = []
for word in text_as_list:
    if word.startswith("a"):
        words_begin_with_a.append(word)
print(words_begin_with_a)

# i = 0
# counter = 0

# while i<len(text_as_list):
#     if text_as_list[i].startswith('a'):
#         counter += 1
#     i += 1
# print(counter)

['about', 'are', 'are', 'alone', 'alien', 'and', 'away', 'all', 'about', 'ago', 'anything', 'and', 'all', 'at', 'and', 'and', 'all', 'about', 'and', 'and', 'and', 'and', 'allow', 'another', 'and', 'and', 'also', 'a', 'are', 'agreement', 'an', 'and', 'are', 'alone', 'arose', 'appear', 'a', 'appeared', 'appearing', 'algae', 'ago', 'ago', 'and', 'about', 'appeared', 'a', 'a', 'appearing', 'available', 'aliens', 'am', 'appear', 'and', 'a', 'and', 'aliens', 'a', 'an', 'any', 'alien', 'are', 'alien', 'at', 'a', 'a', 'an', 'against', 'abduction', 'aliens', 'a', 'are', 'and', 'are', 'an', 'and', 'are', 'along', 'ability', 'and', 'aggressive', 'advantage', 'avoid', 'alone', 'answers', 'am', 'all', 'and', 'answers', 'a', 'a', 'and', 'are']


Answer: 310

#### Create a counter and use a while loop to counter how many words starts with a _u_.

In [28]:
i = 0
counter = 0

while i<len(text_as_list):
    if text_as_list[i].startswith('u'):
        counter += 1
    i += 1
print(counter)

30


Answer: 30

#### Create a counter and use a while loop to counter how many words do NOT start with a _u_.

In [29]:
i = 0
counter = 0
while i<len(text_as_list):
    if not text_as_list[i].startswith('u'):
        counter += 1
    i += 1
print(counter)

877


Answer: 877

#### Write an if statement that checkes if the word "universe" appears in the text. Feel free to google how we test if a word is _in_ a Python list.

In [34]:
if "universe" in text_as_list:
    print("Yes, universe appears in the text.")

Yes, universe appears in the text.


#### Write an if statement that checkes if the word "alliens" appears in the text. If not print "the word does not appear"

In [39]:
if not "aliens" in text_as_list:
    print("the word does not appear")

#### Write an if statement that checkes if the word "alliens" appears in the text. If not, check if the word "International Space Station" appears in the text. If neither is in the text, print "I didn't find the words".

In [41]:
if "aliens" in text_as_list:
    print()
elif "International Space Station" in text_as_list:
    print()
else:
    print("I didn't find the words")

#### Take the text_as_list and create a dictionary named `mydictionary`. The words of the text should be the keys of the dictionary and the index of the word the values of the keys.

Hint: there is a method in Python that adds a counter to an iterable.

In [ ]:
mydictionary = 

#### Create the same dictonary but now index of  words in the text should be the dictionary's keys and the words should be the key's values.

#### Why the lenghts of the two dictionaries are not the same?

#### Create a dictionary `dict_counter` that takes as keys the words of text_as_list. The values of each key should be the number that each words is attested in the list.

Hint: Check the [`count()` method](https://www.w3schools.com/python/ref_list_count.asp).

In [ ]:
print(mydictionary)

#### Without printing the whole dictionary, find how many times the word "question" appears in the list by using your new dictiornary

#### Create a pandas dataframe from the items of the dictionary

#### Get the first row of your dataframe

#### Get the first column of your dataframe

#### By using the `pandas.DataFrame.loc`, find how many times the words galaxies and galaxies appear in the dataframe

#### Write a recursion in a function that prints 5! (i.e., the factorial of 5).

In [ ]:
def fact(n):
    

In [ ]:
print(fact(n))

#### Write a recursion in a function that prints the $n^{th}$ [Fibonacci number](https://en.wikipedia.org/wiki/Fibonacci_number):

1, 1, 2, 3, 5, 8, 13, 21, 34

In [ ]:
def fib(n):
    

#### Which is the $20^{th}$ Fibonacci number?

In [ ]:
fib(20)

_We will try together tail recursion._